opentelemetry-aws/src/lib.rs

```diff
@@ -86,6 +86,108 @@ pub mod trace {
         _private: (),
     }
 
+    #[allow(clippy::result_unit_err)]
+    pub fn deserialize_span_context(value: &str) -> Result<SpanContext, ()> {
+        let parts: Vec<(&str, &str)> = value
+            .split_terminator(';')
+            .filter_map(from_key_value_pair)
+            .collect();
+
+        let mut trace_id: TraceId = TraceId::INVALID;
+        let mut parent_segment_id: SpanId = SpanId::INVALID;
+        let mut sampling_decision = TRACE_FLAG_DEFERRED;
+        let mut kv_vec: Vec<(String, String)> = Vec::with_capacity(parts.len());
+
+        for (key, value) in parts {
+            match key {
+                HEADER_ROOT_KEY => {
+                    let converted_trace_id: Result<TraceId, ()> =
+                        XrayTraceId(value.to_string()).try_into();
+                    match converted_trace_id {
+                        Err(_) => return Err(()),
+                        Ok(parsed) => trace_id = parsed,
+                    }
+                }
+                HEADER_PARENT_KEY => {
+                    parent_segment_id = SpanId::from_hex(value).unwrap_or(SpanId::INVALID)
+                }
+                HEADER_SAMPLED_KEY => {
+                    sampling_decision = match value {
+                        NOT_SAMPLED => TraceFlags::default(),
+                        SAMPLED => TraceFlags::SAMPLED,
+                        REQUESTED_SAMPLE_DECISION => TRACE_FLAG_DEFERRED,
+                        _ => TRACE_FLAG_DEFERRED,
+                    }
+                }
+                _ => kv_vec.push((key.to_ascii_lowercase(), value.to_string())),
+            }
+        }
+
+        match TraceState::from_key_value(kv_vec) {
+            Ok(trace_state) => {
+                if trace_id == TraceId::INVALID {
+                    return Err(());
+                }
+
+                let context: SpanContext = SpanContext::new(
+                    trace_id,
+                    parent_segment_id,
+                    sampling_decision,
+                    true,
+                    trace_state,
+                );
+
+                Ok(context)
+            }
+            Err(trace_state_err) => {
+                global::handle_error(Error::Trace(TraceError::Other(Box::new(trace_state_err))));
+                Err(()) //todo: assign an error type instead of using ()
+            }
+        }
+    }
+
+    pub fn serialize_span_context(span_context: &SpanContext) -> Option<String> {
+        if !span_context.is_valid() {
+            return None;
+        }
+
+        let xray_trace_id: XrayTraceId = span_context.trace_id().into();
+
+        let sampling_decision: &str =
+            if span_context.trace_flags() & TRACE_FLAG_DEFERRED == TRACE_FLAG_DEFERRED {
+                REQUESTED_SAMPLE_DECISION
+            } else if span_context.is_sampled() {
+                SAMPLED
+            } else {
+                NOT_SAMPLED
+            };
+
+        let trace_state_header: String = span_context
```

Comment:
Don't think we need the type annotation here.

Reply:
Fixed